src/Renci.SshNet/Connection/SshConnector.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Renci.SshNet.Channels;
+
+namespace Renci.SshNet.Connection
+{
+    internal sealed class SshConnector : ConnectorBase
+    {
+        private Session _jumpSession;
+        private JumpChannel _jumpChannel;
+
+        public SshConnector(IServiceFactory serviceFactory, ISocketFactory socketFactory)
+            : base(serviceFactory, socketFactory)
+        {
+        }
+
+        public override Socket Connect(IConnectionInfo connectionInfo)
+        {
+            var proxyConnection = connectionInfo.ProxyConnection;
+            if (proxyConnection == null)
+            {
+                throw new ArgumentNullException("connectionInfo.ProxyConnection");
+            }
+
+            if (proxyConnection.GetType() != typeof(ConnectionInfo))
+            {
+                throw new ArgumentException("Expecting connectionInfo to be of type ConnectionInfo");
+            }
```

Comment:
What's going on here? It's a bit unnerving

Reply:
Following the structure in SSH.NET, it would currently be OK to specify `ProxyTypes.Ssh`, but provinding a `ProxyConnectionInfo` instance (intended for HTTP and SOCKS proxy connections) as the `proxyConnection` parameter in the ConnectionInfo constructor. Therefore, I add a check here to ensure the provided proxyConnection can be used to start an SSH session (to the jump server). Still, it may be a better idea to move this check inside the constructor of `ConnectionInfo`.